optuna/samplers/_tpe/sampler.py

```diff
@@ -633,13 +699,53 @@ def _get_observation_pairs(
                 param_value = None
             values[param_name].append(param_value)
 
-    return values, scores
+        if constraints_enabled:
+            assert violations is not None
+            constraint = trial.system_attrs.get(_CONSTRAINTS_KEY)
+            if constraint is None:
+                warnings.warn(
+                    f"Trial {trial.number} does not have constraint values."
+                    " It will be treated as a lower priority than other trials."
+                )
+                violation = float("inf")
+            else:
+                # Violation values of infeasible dimensions are summed up.
+                violation = sum(v for v in constraint if v > 0)
+            violations.append(violation)
+
+    return values, scores, violations
 
 
 def _split_observation_pairs(
     loss_vals: List[Tuple[float, List[float]]],
     n_below: int,
+    violations: Optional[List[float]],
 ) -> Tuple[np.ndarray, np.ndarray]:
+    # When constrains is not None, trials are split into below and above
+    # according to the following rules.
+    # 1. Feasible trials are better than infeasible trials.
+    # 2. Infeasible trials are sorted by sum of how much they violate each constraint.
+    # 3. Feasible trials are sorted by loss_vals.
+    if violations is not None:
+        violation_1d = np.array(violations, dtype=float)
+        idx = violation_1d.argsort()
+        if n_below >= len(idx) or violation_1d[idx[n_below]] > 0:
+            # Below is filled by all feasible trials and trials with smaller violation values.
+            indices_below = idx[:n_below]
+            indices_above = idx[n_below:]
+        else:
+            # All trials in below are feasible.
+            # Feasible trials with smaller loss_vals are selected.
+            (feasible_idx,) = (violation_1d == 0).nonzero()
```

Comment:
```suggestion
            (feasible_idx,) = (violation_1d <= 0).nonzero()
```

Reply:
As we privately discussed, I keep this line because `violation_1d` is guaranteed to be greater than or equal to 0.
https://github.com/optuna/optuna/pull/3506/files#diff-2cc3a4088f95a10e806b68655d375f6f7f5fe2de161a0182c16168e378be8065R712-R713